mtp_api/apps/mtp_auth/serializers.py

```diff
@@ -86,67 +99,103 @@ def get_prisons(self, obj):
             for prison in PrisonUserMapping.objects.get_prison_set_for_user(obj)
         )
 
+    def validate(self, attrs):
+        roles = self.initial_data.get('roles', ())
+        if roles:
+            user = self.context['request'].user
+            managed_roles = {
+                role.name: role
+                for role in Role.objects.get_managed_roles_for_user(user)
+            }
+            try:
+                self.initial_data['roles'] = [managed_roles[role] for role in roles]
+            except KeyError:
+                raise serializers.ValidationError({'roles': 'Invalid role(s)'})
+
+        return super().validate(attrs)
+
     @atomic
     def create(self, validated_data):
         creating_user = self.context['request'].user
+
+        roles = validated_data.pop('roles', ())
         make_user_admin = validated_data.pop('user_admin', False)
         validated_data.pop('is_locked_out', None)
+
+        if not roles:
+            raise serializers.ValidationError({'roles': 'Roles not specified'})
+
         new_user = super().create(validated_data)
 
         password = User.objects.make_random_password(length=10)
         new_user.set_password(password)
+        new_user.save()
 
-        # add application-user mapping
-        client_application = self.context['request'].auth.application
-        usermapping = ApplicationUserMapping.objects.create(
-            user=new_user, application=client_application
-        )
-        new_user.applicationusermapping_set.add(usermapping)
-
-        # add auth groups for application
-        groupmappings = ApplicationGroupMapping.objects.filter(
-            application__client_id=client_application.client_id
-        )
-        for groupmapping in groupmappings:
-            if groupmapping.group in creating_user.groups.all():
-                new_user.groups.add(groupmapping.group)
         if make_user_admin:
             new_user.groups.add(Group.objects.get(name='UserAdmin'))
-        new_user.save()
 
         prisons = PrisonUserMapping.objects.get_prison_set_for_user(creating_user)
         if len(prisons) > 0:
             pu = PrisonUserMapping.objects.create(user=new_user)
             for prison in prisons:
                 pu.prisons.add(prison)
-            pu.save()
+
+        main_application_name = None
+        for role in roles:
+            if main_application_name is None:
+                main_application_name = role.application.name
+            ApplicationUserMapping.objects.create(user=new_user, application=role.application)
```

Comment:
Would this create multiple identical mappings where the user has multiple roles with access to the same application?